plugins/com.google.cloud.tools.eclipse.appengine.libraries.ui/src/com/google/cloud/tools/eclipse/appengine/libraries/ui/CloudLibrariesPage.java

```diff
@@ -19,105 +19,167 @@
 import com.google.cloud.tools.eclipse.appengine.facets.AppEngineStandardFacet;
 import com.google.cloud.tools.eclipse.appengine.libraries.BuildPath;
 import com.google.cloud.tools.eclipse.appengine.libraries.model.Library;
-import com.google.cloud.tools.eclipse.appengine.ui.AppEngineImages;
+import com.google.cloud.tools.eclipse.appengine.libraries.persistence.LibraryClasspathContainerSerializer;
 import com.google.cloud.tools.eclipse.util.MavenUtils;
+import com.google.cloud.tools.eclipse.util.status.StatusUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 import org.eclipse.core.runtime.CoreException;
-import org.eclipse.core.runtime.SubMonitor;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.NullProgressMonitor;
 import org.eclipse.jdt.core.IClasspathEntry;
 import org.eclipse.jdt.core.IJavaProject;
 import org.eclipse.jdt.ui.wizards.IClasspathContainerPage;
 import org.eclipse.jdt.ui.wizards.IClasspathContainerPageExtension;
-import org.eclipse.jdt.ui.wizards.IClasspathContainerPageExtension2;
+import org.eclipse.jface.viewers.StructuredSelection;
 import org.eclipse.jface.wizard.WizardPage;
 import org.eclipse.swt.SWT;
-import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.layout.RowLayout;
 import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Group;
 import org.eclipse.wst.common.project.facet.core.IProjectFacetVersion;
 
 public abstract class CloudLibrariesPage extends WizardPage implements IClasspathContainerPage,
-    IClasspathContainerPageExtension, IClasspathContainerPageExtension2 {
+    IClasspathContainerPageExtension {
 
   private static final Logger logger = Logger.getLogger(CloudLibrariesPage.class.getName());
-  private LibrarySelectorGroup librariesSelector;
+  private final LibraryClasspathContainerSerializer serializer =
+      new LibraryClasspathContainerSerializer();
+
+  /** The library groups to be displayed; pairs of (id, title). */
+  private Map<String, String> groups;
+  /** Selected libraries that are not shown but should be preserved. */
+  private List<Library> hiddenLibraries = Collections.emptyList();
```

Comment:
My guess is that this is the list of libraries that have already been added before opening this page? If so, I think `existingLibraries` or such fits better. `hiddenLibraries` confused me. Also, existing libraries are ticked by default, right? In so, `hiddenLibraries` doesn't sound quite right.

Reply:
This is complicated.  I'll log any libraries that are discarded in setSelectedLibraries().

Reply:
I see, so this means "unrecognized libraries"? Then I think that makes sense.

Reply:
I thought so too, but now I'm not sure.  I can't think of a situation where this will happen and we'd want this behaviour.

Reply:
It may happen if the library definition file changes and some libraries got removed?

Reply:
Oh, there won't be such libraries, as `loadLibraryList` will filter them out since `CloudLibraries.getLibrary(libraryId) == null`?